File: licenses/BuildInfo.php
<?php

class BuildInfo
{
    private $build_info_array;
    private $fields_defined;
    private $multiline_vars;
    private $search_path;
    private $fname;
    private $file_info_array;

    public function __construct($fn)
    {
        $this->build_info_array = array();
        $this->file_info_array = array();
        $this->fields_defined = array("Format-Version", "Files-Pattern",
            "Build-Name", "Theme", "License-Type", "OpenID-Launchpad-Teams",
            "Collect-User-Data", "License-Text");
        $this->multiline_vars = array("License-Text");
        $this->search_path = dirname($fn);
        $this->fname = basename($fn);
        $this->build_info_file = $this->search_path."/BUILD-INFO.txt";
        $data = $this->readFile();
        if (is_array($data)) {
            $this->build_info_array = $this->parseData($data);
            $this->file_info_array = $this->getInfoForFile($this->fname);
        }
    }

    public function readFile()
    {
        $data = array();
        if (is_dir($this->build_info_file) or !is_file($this->build_info_file) or filesize($this->build_info_file) == 0) return false;
        $file = fopen($this->build_info_file, "r") or exit("Unable to open file $this->build_info_file!");
        while(!feof($file)) {
            $line = fgets($file);
            if (trim($line) == "")
                continue;
            $data[] = $line;
        }
        return $data;
    }

    private function getInfoForFile($fname)
    {
        /**
         * Get array of fields for corresponding file
         */
        foreach (array_keys($this->build_info_array) as $key)
            if ($key != 'Format-Version') {
                $files = glob($this->search_path."/".$key);
                foreach ($files as $file)
                    if ($file == $this->search_path."/".$fname)
                        return $this->build_info_array[$key];
            }
        return array();
    }

    public function getFormatVersion()
    {
        if (array_key_exists('Format-Version', $this->build_info_array))
            return $this->build_info_array["Format-Version"];
        else
            return false;
    }

    // Get value of specified field for corresponding file
    public function get($field)
    {
        if (array_key_exists($field, $this->file_info_array))
            return $this->file_info_array[$field];
        else
            return false;
    }

    public function parseLine($line) {
        $values = explode(":", $line, 2);
        if ($values === false || count($values) != 2) {
            throw new InvalidArgumentException("Line is not in the correct format.");
        } else {
            $field = trim($values[0]);
            $value = trim($values[1]);
            if (!$this->isValidField($field)) {
                throw new InvalidArgumentException("Field '$field' not allowed.");
            } else {
                return array($field => $value);
            }
        }
    }

    public function isValidField($field_name) {
        if (in_array($field_name, $this->fields_defined)) {
            return true;
        } else {
            return false;
        }
    }

    public function parseContinuation($lines, &$line_no) {
        $text = '';
        $total_lines = count($lines);
        while ($line_no < $total_lines &&
            strlen($lines[$line_no]) > 0) {
                if ($lines[$line_no][0] == ' ') {
                    $text .= "\n" . substr($lines[$line_no], 1);
                    $line_no++;
                } else {
                    break;
                }
            }
        return $text;
    }

    /**
     * `data` should be array of lines.
     */
    public function parseData($data) {
        if (!is_array($data)) {
            throw new InvalidArgumentException("No array provided.");
        }
        $format_line = array_shift($data);
        $values = $this->parseLine($format_line);
        if (!array_key_exists("Format-Version", $values)) {
            throw new InvalidArgumentException("Data in incorrect format.");
        }
        $result = array("Format-Version" => $values["Format-Version"]);

        $line_no = 0;
        while ($line_no < count($data)) {
            $line = $data[$line_no];
            $values = $this->parseLine($line);
            if (array_key_exists("Files-Pattern", $values)) {
                $line_no++;
                $block = $this->parseBlock($data, $line_no);
                if (is_array($block)) {
                    foreach (explode(",", $values["Files-Pattern"]) as $pattern) {
                        $result[$pattern] = $block;
                    }
                }
            }
        }
        return $result;
    }

    public function parseBlock($data, &$line_no) {
        $result = array();

        if (!is_array($data)) {
            throw new InvalidArgumentException("No array provided.");
        }
        while ($line_no < count($data)) {
            $line = $data[$line_no];
            $values = $this->parseLine($line);
            if (array_key_exists("License-Text", $values)) {
                $text = $values["License-Text"];
                $line_no++;
                $text .= $this->parseContinuation($data, $line_no);
                $result["License-Text"] = $text;
            } elseif (array_key_exists("Files-Pattern", $values)) {
                return $result;
            } else {
                $line_no++;
                $result = array_merge($result, $values);
            }
        }
        return $result;
    }
}

File: license_protected_downloads/buildinfo.py
import os
import fnmatch


class IncorrectDataFormatException(Exception):
        ''' Build-info data is in incorrect format. '''


class BuildInfoBase(object):
    fields_defined = [
        "format-version", "files-pattern", "build-name", "theme",
        "license-type", "auth-groups", "collect-user-data",
        "license-text",
        # Deprecated
        "openid-launchpad-teams",
    ]

    def __init__(self, full_name, content):
        self.fname = os.path.basename(full_name)

        self.index = 0
        self.build_info_array = [{}]

        lines = [x for x in content.split('\n') if x.strip()]
        self.parseData(lines)

        self.file_info_array = self.getInfoForFile()
        self.remove_false_positives()
        self.max_index = len(self.file_info_array)

    # Get value of specified field in block index for
    # corresponding file
    def get(self, field, index=0):
        if index > self.max_index:
            return False
        block = self.file_info_array[index]
        return block.get(field, False)

    def _set(self, key, value):
        """Record set of directives applying to a file pattern
        key: file pattern
        value: list of dicts of field/val pairs"""
        if key in self.build_info_array[self.index]:
            # A repeated key indicates we have found another chunk of
            # build-info
            self.build_info_array[self.index][key].append(value[0])
        else:
            self.build_info_array[self.index][key] = value

    def assertValidField(self, field):
        if field not in self.fields_defined:
            raise IncorrectDataFormatException(
                "Field '%s' not allowed." % field)

    def parseLine(self, line):
        values = line.split(":", 1)
        if len(values) != 2:
            raise IncorrectDataFormatException(
                "'%s': Line is not in the correct format." % line)
        else:
            field = values[0].strip().lower()
            self.assertValidField(field)
            value = values[1].strip()

            # Rename any deprecated field names to new names
            field_renames = {"openid-launchpad-teams": "auth-groups"}
            field = field_renames.get(field, field)
            return {field: value}

    def parseContinuation(self, lines):
        text = ''
        while self.line_no < len(lines) and len(lines[self.line_no]) > 0:
            if lines[self.line_no][0] == ' ':
                text += '\n' + lines[self.line_no][1:]
                self.line_no += 1
            else:
                break
        return text

    def parseBlock(self, lines):
        result = [{}]
        while self.line_no < len(lines):
            line = lines[self.line_no]
            values = self.parseLine(line)
            if 'license-text' in values:
                text = values['license-text']
                self.line_no += 1
                text += self.parseContinuation(lines)
                result[0]['license-text'] = text
            elif 'files-pattern' in values:
                return result
            else:
                self.line_no += 1
                key = values.keys()[0]
                result[0][key] = values[key]
        return result

    def parseData(self, lines):
        if not isinstance(lines, list):
            raise IncorrectDataFormatException("No array provided.")
        if not list:
            raise IncorrectDataFormatException("Empty BUILD-INFO")
        format_line = lines.pop(0)
        values = self.parseLine(format_line)
        if 'format-version' not in values:
            raise IncorrectDataFormatException(
                'Format-Version field not found.')
        self._set("format-version", values["format-version"])

        self.line_no = 0
        while self.line_no < len(lines):
            line = lines[self.line_no]
            values = self.parseLine(line)
            if "files-pattern" in values:
                self.line_no += 1
                block = self.parseBlock(lines)
                if isinstance(block, list):
                    for pattern in values["files-pattern"].split(","):
                            self._set(pattern.strip(), block)

    def getInfoForFile(self):
        for block in self.build_info_array:
            if self.fname in block:
                # File name matches a key directly - don't need to iterate
                # through each using fnmatch to implement globs.
                return block[self.fname]
            for key in block:
                if key != 'format-version':
                    if fnmatch.fnmatch(self.fname, key):
                        return block[key]
        return [{}]

    def remove_false_positives(self):
        open_type = []
        protected_type = []
        index = 0
        for block in self.file_info_array:
            if 'license-type' in block.keys():
                if block["license-type"] == 'open':
                    open_type.append(index)
                if block["license-type"] == 'protected':
                    protected_type.append(index)
                index += 1
            else:
                return False
        if len(protected_type) != 0 and len(open_type) != 0:
            for index in open_type:
                self.file_info_array.pop(index)


class BuildInfo(BuildInfoBase):
    def __init__(self, fn):

        search_path = self.get_search_path(fn)
        build_info_file = os.path.join(search_path, "BUILD-INFO.txt")

        with open(build_info_file, "r") as f:
            super(BuildInfo, self).__init__(fn, f.read())

    @classmethod
    def get_search_path(cls, path):
        "Return BUILD-INFO.txt search path for a given filesystem path."
        if not os.path.isdir(path):
            path = os.path.dirname(path)
        return path

    @classmethod
    def build_info_exists(cls, path):
        "Check if BUILD-INFO.txt exists for a given filesystem path."
        return os.path.exists(
            os.path.join(cls.get_search_path(path), "BUILD-INFO.txt"))

    @classmethod
    def write_from_array(cls, build_info_array, file_path):
        if len(build_info_array[0]):
            with open(file_path, "w") as outfile:
                outfile.write("Format-Version: 0.5\n\n")
                for key in build_info_array[0]:
                    if key != "format-version":
                        outfile.write("Files-Pattern: %s\n" % key)
                        for item in build_info_array[0][key][0]:
                            text = build_info_array[0][key][0][item]
                            if item == "license-text":
                                text = text.replace("\n", "\n ")
                            outfile.write("%s: %s\n" % (item, text))
                        outfile.write("\n")


if __name__ == "__main__":
    import sys
    bi = BuildInfo(sys.argv[1])
    for field in bi.fields_defined:
        print field + " = " + str(bi.get(field))
